dev/Scroller/Scroller.cpp

```diff
@@ -4544,14 +4562,54 @@ void Scroller::SnapPointsVectorChangedHelper(
     MUX_ASSERT(snapPoints);
     MUX_ASSERT(snapPointsSet);
 
-    switch (args.CollectionChange())
+    T insertedItem = nullptr;
+    winrt::CollectionChange collectionChange = args.CollectionChange();
+
+    if (dimension != ScrollerDimension::ZoomFactor)
+    {
+        double viewportSize = dimension == ScrollerDimension::HorizontalScroll ? m_viewportWidth : m_viewportHeight;
+
+        if (collectionChange == winrt::CollectionChange::ItemInserted)
+        {
+            insertedItem = snapPoints.GetAt(args.Index());
+
+            winrt::SnapPointBase winrtSnapPointBase = safe_cast<winrt::SnapPointBase>(insertedItem);
+            SnapPointBase* snapPointBase = winrt::get_self<SnapPointBase>(winrtSnapPointBase);
+
+            // Newly inserted scroll snap point is provided the viewport size, for the case it's not near-aligned.
+            bool snapPointNeedsViewportUpdates = snapPointBase->OnViewportChanged(viewportSize);
+
+            if (dimension == ScrollerDimension::HorizontalScroll)
+            {
+                m_horizontalSnapPointsNeedViewportUpdates |= snapPointNeedsViewportUpdates;
+            }
+            else
+            {
+                m_verticalSnapPointsNeedViewportUpdates |= snapPointNeedsViewportUpdates;
+            }
+        }
+        else if (collectionChange == winrt::CollectionChange::Reset ||
+                 collectionChange == winrt::CollectionChange::ItemChanged)
+        {
+            bool snapPointsNeedViewportUpdates = SnapPointsViewportChangedHelper(snapPoints, viewportSize);
+
+            if (dimension == ScrollerDimension::HorizontalScroll)
+            {
+                m_horizontalSnapPointsNeedViewportUpdates = snapPointsNeedViewportUpdates;
```

Comment:
Is this the only place where this invalidated flag can get cleared? I couldn't find any other place.

Reply:
Yes indeed. This looks correct to me.